### PR TITLE
Add a custom error action

### LINKED ISSUE #615: Add an "exit" link to the main menu
We have a "Start over" option, but we should also have an "exit" option for people who are not logged in but want to delete their information when they finish using an app.

Exit might not be the right word; maybe "delete my answers"

### LINKED ISSUE #620: Create a custom error action
See #232 

Basic features:
1. Show the download screen, with
    * An extra “key” in the ALDocument that’s the fallback version
    * Do the download generation in the background, but show some kind of waiting screen
1. What information would people want on this error screen?
    * How to contact us
    * How to contact the court?
         * Implies a new variable to track the court’s info
     * Ideally, the original form link if it was a paper PDF form
     * Ideally, the document with as much filled in as possible
1. Needs some visual style that makes it clear the user ran into an error
1. ~Will this prevent the built-in feature to email an error report in the config?~
    * Should be something we can do manually, but we need to check
    * Update: Just looked in server.py and the error email is independent of the custom error action.

### PR DESCRIPTION
**Pull request by nonprofittechy**
Creates a custom error page, with an optional link to download the form in progress. The download depends on the `skip undefined` parameter being correctly used on the attachment block.

In addition, this adds an "Exit and delete answers" link (#615) and a link to download the forms in-progress at any time, per discussion on Coding Deep Dive on 11/20/22 and broader discussion in #232

Fix #620 
Fix #615 

image
image

`DACatchAll` is turned on on our development server. If you want to simulate an error, one that I have been trying is triggering a hashing error. E.g., `users[{}]` in the interview order block.

I probably want to keep thinking about this a bit more over the next week or two, but it should be ready to look at. Specifically, I might want to add some features to the GithubFeedbackForm

### CODE REVIEW

File: docassemble/AssemblyLine/data/questions/al_visual.yml [hunk 1]
```diff
@@ -73,15 +73,29 @@ code: |
   if _internal.get('steps') > 1:
     menu_items = al_menu_items
   else:
-    menu_items = al_menu_items[1:] # Skip the "start over" menu item
+    # Skip the "start over", download, and exit and erase menu item
+    if get_config("assembly line", {}).get("enable incomplete downloads"):
+      menu_items = al_menu_items[3:]
+    else:
+      menu_items = al_menu_items[2:]
```
Comment (BryceStevenWilley) [inline docassemble/AssemblyLine/data/questions/al_visual.yml:80]: Do we want this in the config, or do we want it per-interview? I can easily imagine someone wanting it to apply to some interviews but not others.

Comment (nonprofittechy) [reply docassemble/AssemblyLine/data/questions/al_visual.yml:80]: That's fair, I can make it two level

File: docassemble/AssemblyLine/data/questions/al_visual.yml [hunk 2]
```diff
@@ -348,7 +366,119 @@ back button label: |
...
+  % if al_enable_incomplete_downloads:
+  * Try to [download your work in progress](${ url_action("al_error_action_download_screen") })
+  % endif
+  % if MAIN_METADATA.get("original_form") and not MAIN_METADATA.get("original_form").strip() == "None":
+  * [Download a blank "${ all_variables(special='metadata').get("title").strip() }"](${ MAIN_METADATA.get("original_form") } )
+  % endif
+  * Visit the [${ AL_ORGANIZATION_TITLE } home page](${ AL_ORGANIZATION_HOMEPAGE }) and try another interview
+  % if al_show_email_to_user_on_errors and get_config("error notification email"):
+  * [Send us an email](mailto:${ get_config("error notification email") }?subject=User%20error%20report%3A%20${ all_variables(special='metadata').get("title") }%20${ user_info().filename }) (Note: you may not get a reply)
+  % endif
```
Comment (BryceStevenWilley) [inline docassemble/AssemblyLine/data/questions/al_visual.yml:430]: The chance that people actually email us is fairly low, and the auto-emails from errors in interviews have been really useful. Anyway we can just auto-send an email if someone sees this screen, with the same debugging information they would see, or would that require an upstream change

**Review by BryceStevenWilley — COMMENTED** [review]
Haven't had time to really play with it, but it LGTM. I do think that the error emails are really important for production interviews; I don't think this should go to prod until we figure that out. If we want to try this on the dev server for a bit that would be fine, but we might forget that we don't want it on prod if a hotfix comes in, so maybe keep this open until we can figure out the errors emails?

File: docassemble/AssemblyLine/data/questions/al_visual.yml (hunk 2 above)
Comment (nonprofittechy) [reply docassemble/AssemblyLine/data/questions/al_visual.yml:430]: Afaict the error emails still happen when you have a custom error action. Just from looking at the code the emails happen before the error action gets triggered

Comment (nonprofittechy) [reply docassemble/AssemblyLine/data/questions/al_visual.yml:430]: Just gave this a test by turning on notifications on the dev server and I can confirm it works even with the custom error action

**Pull request merged**